### PR TITLE
feat(apollo-smith): implement a graphql test case generator

### LINKED ISSUE #114: apollo-smith: test case generator for apollo-rs crates
## Description
We are maintaining a set of tools for an entire query language, and with that comes a responsibility of having a well tested and spec-compliant code.

One of the ways of doing this is implementing a comprehensive fuzzing system for `apollo-parser` and `apollo-encoder`. I am keen on solving this problem by writing a test case generator. Writing a test case generator early on allows us to preemptively catch spec inconsistencies in our implementation, as well as write a tool that can be reused when we start writing Apollo's composition algorithm in rust, or supporting a core schema implementation.

## Design Details
There are three parts to this work: writing a test case generator, which we will call `apollo-smith`, implementing a fuzzer for `apollo-parser` covered in #59 , and implementing a fuzzer for `apollo-encoder` covered in #60.

### apollo-smith

Before we fuzz the parser and the encoder we should write apollo-smith a test-case generator that fuzzing code can use. [Nick Fitzgerald](https://github.com/fitzgen) has written an amazingly comprehensive [blogpost on writing a test-case generator](https://fitzgeraldnick.com/2020/08/24/writing-a-test-case-generator.html) that I think will be incredibly helpful here. 

`apollo-smith`’s goal is to generate GraphQL programs, both schemas and queries, by sampling from all available possibilities of our grammar. 

Test case generators are usually written like recursive-descent parsers, like Nick points out. The fact that our parser is a recursive-descent parser is handy as we can borrow most of the code and semantics when writing `apollo-smith`. This should make writing this heaps easier!

Tools for us to use when writing this: [`arbitrary` crate](https://docs.rs/arbitrary/1.0.3/arbitrary/), [`libfuzzer-sys` crate] (https://docs.rs/libfuzzer-sys/0.4.2/libfuzzer_sys/), [`cargo fuzz`](https://github.com/rust-fuzz/cargo-fuzz) (cargo fuzz only works on macos and linux systems due to needing llvm, so those of us on windows, ahem @lrlna, will need to boot up wsl).

`wasm-tools` has a quite readable, [wasm-specific implementation](https://github.com/bytecodealliance/wasm-tools/tree/28ee2706878641a4e7d7bdff3c6e532525e47e75/crates/wasm-smith) that we could reference when writing ours. I really like the fact that `wasm-smith` can be used with cargo fuzz, and used in other crates writing wasm implementation. I think if we stick to similar structure, `apollo-smith` has a potential to be used for when we write composition and start supporting core-schemas. This would make our job transitioning to rust-based libraries for some of Apollo’s crucial code a lot more straightforward.

## References

1. [Writing a Test Case Generator for a Programming Language](https://fitzgeraldnick.com/2020/08/24/writing-a-test-case-generator.html), by Nick Fitzgerald.
2. [The original llvm libfuzzer documentation.](https://llvm.org/docs/LibFuzzer.html)
3. [Rust’s libfuzzer-sys cargo subcomand to work with libfuzzer.](https://github.com/rust-fuzz/cargo-fuzz)
4. [Announcing Better Support for Fuzzing with Structured Inputs in Rust](https://fitzgeraldnick.com/2020/01/16/better-support-for-fuzzing-structured-inputs-in-rust.html), by Nick Fitzgerald.
5. [wasm-tools fuzz directory that has examples of fuzzing the wasm parser and encoder.](https://github.com/bytecodealliance/wasm-tools/tree/e67010a741b3e1686ec08ca80e517f8873fd9d4e/fuzz)
6. [wasm-smith, a test case generator for wasm.](https://github.com/bytecodealliance/wasm-tools/tree/e67010a741b3e1686ec08ca80e517f8873fd9d4e/crates/wasm-smith)

### LINKED ISSUE #155: apollo-smith: create the structure of the project
## Description

Related to #114 
The goal is to create the structure of `apollo-smith`. Using the same structure than the `apollo-rs` parser with `todo!` inside functions body. 
Include the dependency of [arbitrary crate](https://docs.rs/arbitrary/1.0.3/arbitrary/) to let us generate arbitrary graphql document.

### LINKED ISSUE #156: apollo-smith: use apollo-encoder
## Description

The best way to finally generate a graphql schema is to use our own `apollo-encoder`. To achieve this we need to find the best solution to use `apollo-encoder` types without having too much impact on the `apollo-encoder` source code. We will need to implement `Arbitrary` trait on types to create arbitrary element of a graphql schema. Here is several solutions we could have regarding the rust orphan rules.

- Create automatically with macros newtypes:

```rust
pub type SchemaDef(apollo_encoder::SchemaDef);
impl Deref for SchemaDef {
    type Target = apollo_encoder::SchemaDef;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for SchemaDef {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl<'a> Arbitrary<'a> for SchemaDef {
    fn arbitrary(u: &mut Unstructured<'a>) -> Result<Self> {
        // My implementation

        Ok(my_schema)
    }
}

// Use a macro to create these types automatically
```

- Re-creating our own types and implement `Into` to convert to an `apollo-encoder` type

```rust
#[derive(Debug, Clone)]
pub struct SchemaDef {
    // Description may be a String.
    description: StringValue,
    // The vector of fields in a schema to represent root operation type
    // definition.
    query: Option<String>,
    mutation: Option<String>,
    subscription: Option<String>,
}

impl Into<apollo_encoder::SchemaDef> for SchemaDef {
    fn into(schema: Self) -> apollo_encoder::SchemaDef {
        apollo_encoder::SchemaDef {
            description: schema.description,
            query: schema.query,
            mutation: schema.mutation,
            subscription: schema.subscription,
        }
    }
}

impl<'a> Arbitrary<'a> for SchemaDef {
    fn arbitrary(u: &mut Unstructured<'a>) -> Result<Self> {
        // My implementation

        Ok(my_schema)
    }
}
```

- Implement directly `Arbitrary` trait in the `apollo-encoder` with a feature flag

```rust
#[cfg(feature = "fuzzing")]
impl<'a> Arbitrary<'a> for SchemaDef {
    fn arbitrary(u: &mut Unstructured<'a>) -> Result<Self> {
        // My implementation

        Ok(my_schema)
    }
}
```

The latter is not the best solution because there will be a lot of additions in the `apollo-encoder` IMHO.

My favorite and cleaner solution would be to use the first solution based on the newtype pattern.

### PR DESCRIPTION
**Pull request by bnjjj**
close #114 
close #155 
close #156

### CODE REVIEW

**Review by lrlna — CHANGES_REQUESTED** [review]
`apollo-smith` is looking really really good. Super excited to land this!

There are two main things I noticed in this current state:
- I don't see "ignored token" implementation in the test case generator. I bet we are going to find a whole lot of bugs while fuzzing once this is added.
- Similar to a comment I made in the encoder PR, Operation Definition can have a shorthand, and the test case generator should account for it. The shorthand is only valid when it's the only Operation Definition in a given Document.

File: fuzz/fuzz_targets/parser.rs [hunk 1]
```diff
@@ -1,27 +1,44 @@
...
+        let errors = tree
+            .errors()
+            .map(|err| err.message())
+            .collect::<Vec<&str>>()
+            .join("\n");
+        debug!("======= DOCUMENT =======");
+        debug!("{}", doc_generated);
+        debug!("========================");
+        debug!("Parser error ============= \n{:?}", errors);
+        debug!("========================");
```
[the remark this answers is not in the document]
Comment (bnjjj) [reply fuzz/fuzz_targets/parser.rs:39]: Oh yes you're talking about `log_gql_doc` function. I forgot to add it when there is an error. It's done now. And about the `debug!` logging it's just useful when you're debugging or developing `apollo-smith` to see what you're generating.

File: crates/apollo-smith/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            interface_type_defs: Vec::new(),
+            enum_type_defs: Vec::new(),
+            schema_defs: Vec::new(),
+            directive_defs: Vec::new(),
+            operation_defs: Vec::new(),
+            fragment_defs: Vec::new(),
+            scalar_type_defs: Vec::new(),
+            union_type_defs: Vec::new(),
+            input_object_type_defs: Vec::new(),
+        };
```
[the remark this answers is not in the document]
Comment (bnjjj) [reply crates/apollo-smith/src/lib.rs:86]: Created https://github.com/apollographql/apollo-rs/issues/173

**Review by lrlna — APPROVED** [review]
Exciting!

**Pull request merged**